makefile

```diff
@@ -394,6 +394,12 @@ BIGENDIAN := 1
 endif
 endif # BIGENDIAN
 
+# Work around an issue with long doubles on ppc64 (#3157)
+ifneq (,$(findstring ppc64,$(UNAME)))
+ARCHOPTS_C += -mlong-double-64
+ARCHOPTS_CXX += -mlong-double-64
+endif
```

Comment:
How safe is this?  Does Linux provide a C runtime for 64-bit long double ABI, or alternate versions of the entry points for printf-like functions?  If it doesn't, this will potentially cause stack overruns.

Reply:
Near as I can determine the use of 128-bit long double is optional, not mandatory, and libc is designed to swing both ways. See, for example, https://gcc.gnu.org/ml/gcc-patches/2006-01/msg01958.html

Reply:
(I should say, on ppc64/ppc64le, which is the only place this patch would do anything.)

Reply:
That patch doesn't allow you to build with against glibc 2.4 or late with `-mlong-double-64` - it only provides backwards compatibility for programs built against older versions of glibc.  It assumes that if you're building against a later version of glibc, you're using 128-bit long doubles.

Reply:
I'm not sure what you're implying: that MAME on ppc64 doesn't work with 64-bit long doubles? I can definitely assure you that it does, I was playing Alien Syndrome in it earlier and using the Tomy Tutor emulator.

If there are portions in mame that you are concerned may break with this change, I couldn't find any personally, but I'm willing to fix any suspect areas on ppc64 if you point me in the right direction (with the clear understanding they're not going to occur anywhere else *but* ppc64). If you're implying that the compiler should be fixed in lieu of this change, however, given how long the gcc PR has rotted I don't see that happening anytime soon and that would definitely be beyond my personal capability.

Reply:
I'm saying that sooner or later it's going to blow up when something in MAME passes a long double to a printf-like function (e.g. osd_printf_warning or something).  You can only safely use 64-bit long double if you can guarantee a long double will never be passed to the C runtime, and that's definitely not the case for MAME.  You can pass those flags on your own make command line if you want, but you get to keep the pieces when it breaks.  I don't want want MAME to build in a way that's known to be unsafe out-of-the-box.

It's a GCC bug, and they need to fix it.  It's more important than ever with modern C++ because this bug means that user-defined literal suffixes (introduced in C++11) can't be used for floating-point numbers on 64-bit PowerPC targets with GCC.

Reply:
Please be explicit. Are you asking me to fix the printfs in MAME that might get a long double (I'm willing to find these), or are you saying you won't take this under any circumstances until whenever `gcc` is fixed, which may be never?

If the former, I'm game. If the latter, I think this sucks, because the platform will probably remain unbuildable for an extended period of time (possibly permanently) and I substantially dispute your conclusion which isn't supported by any documentation I can find. More to the point, you're making this statement as if you actually support ppc64le, which the very existence of this PR shows you don't. I'm trying not to be pissy, but I think your call here is at best misguided and at worst capricious.

Assuming the latter, which I suspect, it's your call and your project, and I won't submit any further PR on this issue in that case. You can do with it as you please. The `bx` fix can be dealt with as upstream wants and you can import it at your leisure since it won't fix the platform by itself.

Reply:
And, if the former, I'll start on that as soon as I get back from vacation.

Reply:
No, I'm saying that with its scope, MAME can't guarantee it won't pass a long double to the C runtime library, and hence can't be built safely with non-standard ABI options like this.  There's no way we're going to enforce a rule that developers need to be careful to avoid ever passing a long double to the C runtime library and to look for this when reviewing submissions.  It just isn't practical.  We don't support building with non-standard ABI options on x86 either.

The GCC people need to fix their compiler.  A significant piece of language functionality doesn't work, and they can't expect projects to just avoid that because they're unwilling to implement it.

In the mean time, you can add additional flags when you build your own MAME binaries, but you need to be aware that this is very much unsupported by us.  It's acceptable for mainline.